Fixes for a few minor bugs in the Sodium GUI

bytzo opened the pull request:
### Overview

This pull request contains fixes for 6 minor bugs in Sodium's GUI. See each commit and corresponding commit message for more detail, or read on below with some visual examples.

### Fixes

#### Prevent slider thumb from rendering outside of slider area (c8d5645)

The slider thumb in the slider control can currently render outside of the slider area when the value of the slider is outside of its maximum range. While this shouldn't happen during normal use, it can occur if the game options file is manually edited, which is common in speedrunning when the player wants maximum brightness (maximum brightness is 12.0 when the slider goes up to 1.0). Instead of having the slider thumb render outside of the slider area when situations like this occur, we can limit its position to be within the minimum and maximum values of the slider.

#### Remove offset from mouse when setting value of slider control (afb3158)

Currently, when adjusting a value using the slider control, the range of the mouse is shortened by 4 pixels on each side. This means that if the mouse is four pixels inside from the left of the slider, the value is incorrectly set as the minimum. 

An example is shown in the clip below. The top slider shows the current behavior, with the range of the mouse shortened by 4 pixels on each side. The bottom slider shows the proposed fix, which is to remove the 4 pixel offset.

https://user-images.githubusercontent.com/69266322/110059787-c81e1080-7d19-11eb-8056-cb26ddcb888f.mov

Fixing this issue makes the positioning of the slider control and the set value the same, preventing "drift" of the slider thumb when using it.

#### Allow controls to reset themselves when modified options are undone (6bd054f and 4fda9e5)

Currently, control elements keep their fields when modified options are undone. This causes issues when changing a control element and pressing "Undo," because fields such as the slider position of the slider control or the current index of the cycle control are not reset.

An example is shown in the clip below. The clip shows the issue when the control element (in this case the render distance slider) does not reset after it is modified and the "Undo" button is pressed. In the clip, the render distance is modified from 12 to 24 before resetting, but the slider still displays the value of 24 instead of 12.

https://user-images.githubusercontent.com/69266322/110069147-10dec500-7d2c-11eb-999e-ece12098846d.mov

To solve these issues, the control elements will now fetch the updated values when necessary, either on hover (in the case of the slider) or on click (in the case of the cycle).

#### Compare current and modified option values using `equals()` (1970cc2)

The current and modified option values in the Sodium options need to be compared to see if the value is changed or not. Previously, this was done using the `==` operator. However, since the current and modified options are wrapper types, it will compare them as references, making them unequal when their values are the same. This issue does not happen in a majority of the Sodium options, since Java may cache the objects as values. To solve this issue, the current and modified options can be compared using `equals()`, which will compare by value instead. See [this article](https://programming.guide/java/boxed-values-equality.html) about this.

#### Prevent hidden button widgets from being pressed (24158b1)

Currently, button widgets that are not visible can be clicked on, since only the enabled state of the button is considered if an action should be taken or not. To fix this issue, button widgets should also consider their own visibility state when taking action after a click.

#### Allow for more precise widget rendering (69e031b)

Currently, the rendering of rectangles and quads in the Sodium GUI is defined using integers. This restricts the position of widgets to each in-game pixel, meaning using a scaled GUI scale causes the rendered position to be incorrect in some scenarios. For example, a user on a HiDPI display using a GUI scale of 8 would not see the brightness slider thumb move when setting it from 0 to 1, because it must be aligned to every in-game pixel (in this scenario, each in-game pixel is 8 real pixels on the display, because of a GUI scale of 8). To solve this, the rendering of rectangles and quads should use doubles to define position, allowing for greater positioning precision when needed.

bytzo committed:
fix: Prevent slider thumb from rendering outside of slider area
The slider thumb in the slider control previously could render outside of the slider area when the value of the slider was outside of its maximum range. While this shouldn't happen during normal use, it could occur if the game options file was manually edited, which is common in speedrunning when the player wants maximum brightness. Instead of having the slider thumb render outside of the slider area when situations like this occur, we can limit its position to be within the minimum and maximum values of the slider.

bytzo committed:
fix: Remove offset from mouse when setting value of slider control
Previously, when adjusting a value using the slider control, the range of the mouse was shortened by 4 pixels on each side. This meant that if the mouse was four pixels inside from the left of the slider, the value would incorrectly set as the minimum. This offset has been removed to make the positioning of the slider control and the set value the same, preventing "drift" of the slider thumb.

bytzo committed:
fix: Allow controls to reset themselves when modified options are undone
Previously, control elements kept their fields when modified options were undone. This caused issues when changing a control element and pressing "Undo," because fields such as the slider position of the slider control or the current index of the cycle control were not reset. To solve these issues, the control elements that need to be reset on undo now implement a reset interface, which allows them to reset these fields.

FlashyReese commented on src/main/java/me/jellysquid/mods/sodium/client/gui/options/control/SliderControl.java:
```diff
@@ -157,6 +158,11 @@ public boolean mouseDragged(double mouseX, double mouseY, int button, double del
+
+        @Override
+        public void reset() {
+            this.thumbPosition = this.getThumbPositionForValue(option.getValue());
```
I have just come to realize how needless that interface was, you can simply call whatever you need to reset on the render method. In `SliderControl`'s case I would place it in `renderSlider` method. Same thing applies for `CyclingControl`.

bytzo replied:
I have removed the interface in 4fda9e5ac2e541358b45c48416e3c3f273e530ae.

bytzo committed:
change: Remove resettable interface
Removes the resettable interface, because we can update the values through existing methods in each control instead.

Co-authored-by: Yao Chung Hu <30311066+FlashyReese@users.noreply.github.com>

CalXee commented:
Entity distance, if you set it to anything that's not 100% or 125% causes some issues. Namely, if you for example set it to 500%, then slide it to another value then back to 500%, it will still prompt you to apply the entity distance changes despite you didn't actually change anything. I hope I explained that well.

bytzo committed:
fix: Compare current and modified option values using equals()
The current and modified option values in the Sodium options need to be compared to see if the value is changed or not. Previously, this was done using the == operator. However, since the current and modified options are wrapper types, it will compare them as references, making them inequal when their values are the same. This issue does not happen in a majority of the Sodium options, since Java may cache the objects as values. To solve this issue, the current and modified options can be compared using equals(), which will compare by value instead.

bytzo commented:
> Entity distance, if you set it to anything that's not 100% or 125% causes some issues. Namely, if you for example set it to 500%, then slide it to another value then back to 500%, it will still prompt you to apply the entity distance changes despite you didn't actually change anything. I hope I explained that well. This isn't an issue specific to this commit, I'm just saying this in case you guys are unaware.

I added commit 1970cc2074f12c1236d4adab8172ded78279eb34 to fix this. It turns out to be an issue where Sodium compares two wrapper types as references, not values. And since these can be cached as values, it is not seen on sliders that do not exceed 127. I found [this article](https://programming.guide/java/boxed-values-equality.html) when trying to understand the issue, it explains what is happening well.

bytzo committed:
fix: Prevent hidden button widgets from being pressed
Previously, button widgets that were not visible could be clicked on, since only the enabled state of the button was considered if action should be taken or not. To fix this issue, button widgets will also consider their own visibility state when taking action after a click.

bytzo committed:
fix: Change spelling of "visibel" to "visible"

bytzo committed:
fix: Allow for more precise widget rendering
Previously, the rendering of rectangles and quads in the Sodium GUI was defined using integers. This restricted the position of widgets to each in-game pixel, meaning using a scaled GUI scale would cause the rendered position to be incorrect in some scenarios. For example, a user on a HiDPI display using a GUI scale of 8 would not see the brightness slider thumb move when setting it from 0 to 1, because it must be aligned to every in-game pixel (in this scenario, each in-game pixel is 8 real pixels on the display, because of a GUI scale of 8). To solve this, the rendering of rectangles and quads now uses doubles to define position, allowing for greater positioning precision when needed.

jellysquid3 commented:
Everything here looks good to me. Apologies for the delay on merging this, and thanks for your extensive documentation on the issues + your solutions.
